qa/tasks/cephfs/test_volumes.py

```diff
@@ -1290,6 +1290,150 @@ def test_authorize_deauthorize_subvolume(self):
+
+        # create group
+        self._fs_cmd("subvolumegroup", "create", self.volname, group)
+
+        # mounts[0] and mounts[1] would be used as guests to mount the volumes/shares.
+        for i in range(0, 2):
+            self.mounts[i].umount_wait()
+        guest_mounts = (self.mounts[0], self.mounts[1])
+        auth_id = "guest"
+        guestclient_1 = {
+            "auth_id": auth_id,
+        }
+
+        # Create two subvolumes. Authorize 'guest' auth ID to mount the two
+        # subvolumes. Mount the two subvolumes. Write data to the volumes.
+        for i in range(2):
+            # Create subvolume.
+            self._fs_cmd("subvolume", "create", self.volname, subvolumes[i], "--group_name", group)
+
+            # authorize guest authID read-write access to subvolume
+            key = self._fs_cmd("subvolume", "authorize", self.volname, subvolumes[i], guestclient_1["auth_id"],
+                               "--group_name", group)
+
+            mount_path = self._fs_cmd("subvolume", "getpath", self.volname, subvolumes[i],
+                                      "--group_name", group).rstrip()
+            # configure credentials for guest client
+            self._configure_guest_auth(guest_mounts[i], auth_id, key)
+
+            # mount the subvolume, and write to it
+            guest_mounts[i].mount(cephfs_mntpt=mount_path)
+            guest_mounts[i].write_n_mb("data.bin", 1)
+
+        # Evict client, guest_mounts[0], using auth ID 'guest' and has mounted
+        # one volume.
+        self._fs_cmd("subvolume", "evict", self.volname, subvolumes[0], auth_id, "--group_name", group)
+
+        # Evicted guest client, guest_mounts[0], should not be able to do
+        # anymore metadata ops.  It should start failing all operations
+        # when it sees that its own address is in the blocklist.
+        try:
+            guest_mounts[0].write_n_mb("rogue.bin", 1)
+        except CommandFailedError:
+            pass
+        else:
+            raise RuntimeError("post-eviction write should have failed!")
+
+        # The blocklisted guest client should now be unmountable
+        guest_mounts[0].umount_wait()
+
+        # Since '--use-subvol-path' is not used during eviction, all subvolume clients with 'guest'
+        # auth-id is evicted. This mount too should fail.
+        try:
+            guest_mounts[1].write_n_mb("rogue.bin", 1)
+        except CommandFailedError:
+            pass
+        else:
+            raise RuntimeError("post-eviction write should have failed!")
+
+        # The blocklisted guest client should now be unmountable
+        guest_mounts[1].umount_wait()
+
+        # Cleanup.
+        for i in range(2):
+            self._fs_cmd("subvolume", "deauthorize", self.volname, subvolumes[i], auth_id, "--group_name", group)
+            self._fs_cmd("subvolume", "rm", self.volname, subvolumes[i], "--group_name", group)
+        self._fs_cmd("subvolumegroup", "rm", self.volname, group)
+
```

Comment:
I just ran these tests and got the failures.

```
2021-01-13T11:36:25.967+0000 7f00897fa700 -1 client.4952 failed to remount (to trim kernel dentries): return code = 1
ceph-fuse[27203]: fuse finished with error 0 and tester_r 1
2021-01-13 11:36:26,732.732 INFO:tasks.cephfs.mount:Cleaning up mount local
rmdir: failed to remove '/tmp/tmp5jz97dsj/mnt.1': No such file or directory
2021-01-13 11:36:29,849.849 INFO:__main__:Stopped test: test_subvolume_evict_client_use_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes) in 48.097837s
2021-01-13 11:36:29,849.849 INFO:__main__:test_subvolume_evict_client_use_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes) ... ERROR
2021-01-13 11:36:29,850.850 INFO:__main__:
2021-01-13 11:36:29,850.850 INFO:__main__:======================================================================
2021-01-13 11:36:29,850.850 INFO:__main__:ERROR: test_subvolume_evict_client_use_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes)
2021-01-13 11:36:29,850.850 INFO:__main__:----------------------------------------------------------------------
2021-01-13 11:36:29,850.850 INFO:__main__:Traceback (most recent call last):
2021-01-13 11:36:29,850.850 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/test_volumes.py", line 1331, in test_subvolume_evict_client_use_subvol_path
2021-01-13 11:36:29,850.850 INFO:__main__:    guest_mounts[i].write_n_mb("data.bin", 1)
2021-01-13 11:36:29,850.850 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 965, in write_n_mb
2021-01-13 11:36:29,850.850 INFO:__main__:    ], wait=wait)
2021-01-13 11:36:29,851.851 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 680, in run_shell
2021-01-13 11:36:29,851.851 INFO:__main__:    return self.client_remote.run(args=args, cwd=cwd, timeout=timeout, stdout=stdout, stderr=stderr, **kwargs)
2021-01-13 11:36:29,851.851 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 399, in run
2021-01-13 11:36:29,851.851 INFO:__main__:    return self._do_run(**kwargs)
2021-01-13 11:36:29,851.851 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 467, in _do_run
2021-01-13 11:36:29,851.851 INFO:__main__:    proc.wait()
2021-01-13 11:36:29,851.851 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 213, in wait
2021-01-13 11:36:29,851.851 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
2021-01-13 11:36:29,851.851 INFO:__main__:teuthology.exceptions.CommandFailedError: Command failed with status 1: ['dd', 'if=/dev/urandom', 'of=data.bin', 'bs=1M', 'conv=fdatasync', 'count=1', 'seek=0']
2021-01-13 11:36:29,852.852 INFO:__main__:
Cannot find device "ceph-brx"
2021-01-13 11:36:29,872.872 INFO:__main__:
2021-01-13 11:36:29,872.872 INFO:__main__:----------------------------------------------------------------------
2021-01-13 11:36:29,872.872 INFO:__main__:Ran 1 test in 48.099s
2021-01-13 11:36:29,872.872 INFO:__main__:
2021-01-13 11:36:29,872.872 INFO:__main__:FAILED (errors=1)
2021-01-13 11:36:29,872.872 INFO:__main__:
2021-01-13 11:36:29,873.873 INFO:__main__:
2021-01-13 11:36:29,873.873 INFO:__main__:======================================================================
2021-01-13 11:36:29,873.873 INFO:__main__:ERROR: test_subvolume_evict_client_use_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes)
2021-01-13 11:36:29,873.873 INFO:__main__:----------------------------------------------------------------------
2021-01-13 11:36:29,873.873 INFO:__main__:Traceback (most recent call last):
2021-01-13 11:36:29,873.873 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/test_volumes.py", line 1331, in test_subvolume_evict_client_use_subvol_path
2021-01-13 11:36:29,873.873 INFO:__main__:    guest_mounts[i].write_n_mb("data.bin", 1)
2021-01-13 11:36:29,873.873 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 965, in write_n_mb
2021-01-13 11:36:29,873.873 INFO:__main__:    ], wait=wait)
2021-01-13 11:36:29,873.873 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 680, in run_shell
2021-01-13 11:36:29,874.874 INFO:__main__:    return self.client_remote.run(args=args, cwd=cwd, timeout=timeout, stdout=stdout, stderr=stderr, **kwargs)
2021-01-13 11:36:29,874.874 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 399, in run
2021-01-13 11:36:29,874.874 INFO:__main__:    return self._do_run(**kwargs)
2021-01-13 11:36:29,874.874 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 467, in _do_run
2021-01-13 11:36:29,874.874 INFO:__main__:    proc.wait()
2021-01-13 11:36:29,874.874 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 213, in wait
2021-01-13 11:36:29,874.874 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
2021-01-13 11:36:29,874.874 INFO:__main__:teuthology.exceptions.CommandFailedError: Command failed with status 1: ['dd', 'if=/dev/urandom', 'of=data.bin', 'bs=1M', 'conv=fdatasync', 'count=1', 'seek=0']
2021-01-13 11:36:29,874.874 INFO:__main__:
2021-01-13 11:36:29,899.899 INFO:__main__:
2021-01-13 11:36:29,900.900 INFO:__main__:
```

```
ceph-fuse[28562]: starting ceph client
ceph-fuse[28562]: starting fuse
mount: only root can use "--options" option
2021-01-13T11:37:28.407+0000 7f8d3d7fa700 -1 client.5236 failed to remount (to trim kernel dentries): return code = 1
ceph-fuse[28562]: fuse finished with error 0 and tester_r 1
2021-01-13 11:37:29,290.290 INFO:tasks.cephfs.mount:Cleaning up mount local
rmdir: failed to remove '/tmp/tmp6pwy8ost/mnt.1': No such file or directory
2021-01-13 11:37:32,805.805 INFO:__main__:Stopped test: test_subvolume_evict_client_no_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes) in 50.461764s
2021-01-13 11:37:32,806.806 INFO:__main__:test_subvolume_evict_client_no_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes) ... ERROR
2021-01-13 11:37:32,806.806 INFO:__main__:
2021-01-13 11:37:32,806.806 INFO:__main__:======================================================================
2021-01-13 11:37:32,806.806 INFO:__main__:ERROR: test_subvolume_evict_client_no_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes)
2021-01-13 11:37:32,806.806 INFO:__main__:----------------------------------------------------------------------
2021-01-13 11:37:32,806.806 INFO:__main__:Traceback (most recent call last):
2021-01-13 11:37:32,807.807 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/test_volumes.py", line 1400, in test_subvolume_evict_client_no_subvol_path
2021-01-13 11:37:32,807.807 INFO:__main__:    guest_mounts[i].write_n_mb("data.bin", 1)
2021-01-13 11:37:32,807.807 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 965, in write_n_mb
2021-01-13 11:37:32,807.807 INFO:__main__:    ], wait=wait)
2021-01-13 11:37:32,807.807 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 680, in run_shell
2021-01-13 11:37:32,807.807 INFO:__main__:    return self.client_remote.run(args=args, cwd=cwd, timeout=timeout, stdout=stdout, stderr=stderr, **kwargs)
2021-01-13 11:37:32,807.807 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 399, in run
2021-01-13 11:37:32,807.807 INFO:__main__:    return self._do_run(**kwargs)
2021-01-13 11:37:32,808.808 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 467, in _do_run
2021-01-13 11:37:32,808.808 INFO:__main__:    proc.wait()
2021-01-13 11:37:32,808.808 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 213, in wait
2021-01-13 11:37:32,808.808 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
2021-01-13 11:37:32,808.808 INFO:__main__:teuthology.exceptions.CommandFailedError: Command failed with status 1: ['dd', 'if=/dev/urandom', 'of=data.bin', 'bs=1M', 'conv=fdatasync', 'count=1', 'seek=0']
2021-01-13 11:37:32,808.808 INFO:__main__:
Cannot find device "ceph-brx"
2021-01-13 11:37:32,827.827 INFO:__main__:
2021-01-13 11:37:32,828.828 INFO:__main__:----------------------------------------------------------------------
2021-01-13 11:37:32,828.828 INFO:__main__:Ran 1 test in 50.463s
2021-01-13 11:37:32,828.828 INFO:__main__:
2021-01-13 11:37:32,828.828 INFO:__main__:FAILED (errors=1)
2021-01-13 11:37:32,828.828 INFO:__main__:
2021-01-13 11:37:32,828.828 INFO:__main__:
2021-01-13 11:37:32,828.828 INFO:__main__:======================================================================
2021-01-13 11:37:32,828.828 INFO:__main__:ERROR: test_subvolume_evict_client_no_subvol_path (tasks.cephfs.test_volumes.TestSubvolumes)
2021-01-13 11:37:32,829.829 INFO:__main__:----------------------------------------------------------------------
2021-01-13 11:37:32,829.829 INFO:__main__:Traceback (most recent call last):
2021-01-13 11:37:32,829.829 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/test_volumes.py", line 1400, in test_subvolume_evict_client_no_subvol_path
2021-01-13 11:37:32,829.829 INFO:__main__:    guest_mounts[i].write_n_mb("data.bin", 1)
2021-01-13 11:37:32,829.829 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 965, in write_n_mb
2021-01-13 11:37:32,829.829 INFO:__main__:    ], wait=wait)
2021-01-13 11:37:32,829.829 INFO:__main__:  File "/home/jcollin/ceph/qa/tasks/cephfs/mount.py", line 680, in run_shell
2021-01-13 11:37:32,829.829 INFO:__main__:    return self.client_remote.run(args=args, cwd=cwd, timeout=timeout, stdout=stdout, stderr=stderr, **kwargs)
2021-01-13 11:37:32,829.829 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 399, in run
2021-01-13 11:37:32,830.830 INFO:__main__:    return self._do_run(**kwargs)
2021-01-13 11:37:32,830.830 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 467, in _do_run
2021-01-13 11:37:32,830.830 INFO:__main__:    proc.wait()
2021-01-13 11:37:32,830.830 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 213, in wait
2021-01-13 11:37:32,830.830 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
2021-01-13 11:37:32,830.830 INFO:__main__:teuthology.exceptions.CommandFailedError: Command failed with status 1: ['dd', 'if=/dev/urandom', 'of=data.bin', 'bs=1M', 'conv=fdatasync', 'count=1', 'seek=0']
2021-01-13 11:37:32,830.830 INFO:__main__:
2021-01-13 11:37:32,855.855 INFO:__main__:
2021-01-13 11:37:32,855.855 INFO:__main__:
```

Reply:
It's working for me locally. Could you check the output of "ceph auth get client.guest" and check that subvolpath has been added correctly?

Reply:
@joscollin please try it now. I have refreshed the PR